### PR TITLE
Add Gameboard class and few utils

### PR DESCRIPTION
**Pull request by dariannax3**
Issue: #2

### CODE REVIEW

File: tests/GameboardTest.cpp [hunk 1]
```diff
@@ -20,4 +21,133 @@ TEST(RandomGeneratorTest, getRandomTest) {
...
+protected:
+    Gameboard gameboard;
+
+    void printBoard() { 
+        StdViewer viewer;
+        viewer.drawBoard(gameboard);
+    }
+
+    bool isRangeCorrect(int row_start, int row_end, int column_start, int column_end){
+        return row_start > 7 || row_start < 0 || row_end > 7 || row_end < 0 || column_start > 7 || column_start < 0 || column_end > 7 || column_end < 0;
```
Comment (Rosein) [inline tests/GameboardTest.cpp:36]: Magic numbers. Please replace them with something meaningful.

[https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es45-avoid-magic-constants-use-symbolic-constants](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es45-avoid-magic-constants-use-symbolic-constants)

File: tests/GameboardTest.cpp [hunk 2]
```diff
@@ -20,4 +21,133 @@ TEST(RandomGeneratorTest, getRandomTest) {
...
+
+    checkVisibilityOfSpecificRowsAndFields(0, 1, 0, 7, Visibility::covered);
+    checkVisibilityOfSpecificRowsAndFields(2, 2, 0, 4, Visibility::covered);
+    checkVisibilityOfSpecificRowsAndFields(2, 2, 5, 5, Visibility::uncovered);
+    checkVisibilityOfSpecificRowsAndFields(2, 2, 6, 7, Visibility::covered);
+    checkVisibilityOfSpecificRowsAndFields(4, 7, 0, 7, Visibility::covered);
+
+    EXPECT_TRUE(gameboard.getFieldAt(1, 4).flagability == Flagability::marked);
+
 }
```
Comment (Rosein) [inline tests/GameboardTest.cpp:153]: Add new line at the end of file.

File: tests/GameboardTest.cpp [hunk 3]
```diff
@@ -20,4 +21,133 @@ TEST(RandomGeneratorTest, getRandomTest) {
...
+
+    EXPECT_EQ(gameboard.countAdjacentBombAt(0, 0), 1);
+    EXPECT_EQ(gameboard.countAdjacentBombAt(0, 1), 1);  
+
+    checkVisibilityOfSpecificRowsAndFields(0, 1, 0, 1, Visibility::covered);
+    checkVisibilityOfSpecificRowsAndFields(0, 2, 2, 7, Visibility::uncovered);
+    checkVisibilityOfSpecificRowsAndFields(2, 7, 0, 7, Visibility::uncovered);
+}
+
+TEST_F(GameboardTestFixture, givenBoardWithEightFlagsFormingSquareWithUncoverFieldInsideWhenClickInUncoverFIeldsThenShouldUncoverOnlyThatField){
```
Comment (Rosein) [inline tests/GameboardTest.cpp:141]: **g**iven

File: inc/Gameboard.hpp [hunk 4]
```diff
@@ -1,11 +1,45 @@
-#pragma once
+﻿#pragma once
 #include <memory>
 #include "RandomGeneratorI.hpp"
 
 using RandomGeneratorPtr = std::shared_ptr<RandomGeneratorI>;
 
+enum class Visibility { covered,
+                        uncovered };
+// don't ask
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Gameboard.hpp:9]: done

File: tests/GameboardTest.cpp (hunk 2 above)
Comment (dariannax3) [reply tests/GameboardTest.cpp:163]: done

File: inc/Viewer.hpp [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿#pragma once
+#include "Gameboard.hpp"
+#include <chrono>
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Viewer.hpp]: done

File: inc/Gameboard.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Gameboard.hpp:9]: done

File: inc/Gameboard.hpp [hunk 6]
```diff
@@ -1,11 +1,45 @@
...
+                       mined };
+
+// don't ask again
+enum class Flagability { unmarked,
+                       marked };
+
+struct Field {
+    Visibility visibility = Visibility::covered;
+    Bombility bombility = Bombility::empty;
+    Flagability flagability = Flagability::unmarked;
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Gameboard.hpp]: done

File: inc/Gameboard.hpp [hunk 7]
```diff
@@ -1,11 +1,45 @@
...
 
 using RandomGeneratorPtr = std::shared_ptr<RandomGeneratorI>;
 
+enum class Visibility { covered,
+                        uncovered };
+// don't ask
+enum class Bombility { empty,
+                       mined };
+
+// don't ask again
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Gameboard.hpp:12]: done

File: inc/Gameboard.hpp [hunk 8]
```diff
@@ -1,11 +1,45 @@
...
+    void uncoverAllFields();
+    void uncoverOneField(const int x, const int y);
+    void flagField(const int x, const int y);
+    const Field& getFieldAt(const int row, const int column) const;
+    Field& getFieldAt(const int row, const int column);
+
 private:
+    void uncoverOneFieldIfPossible(const int x, const int y);
+
     std::shared_ptr<RandomGeneratorI> generator;
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply inc/Gameboard.hpp]: done

File: src/Viewer.cpp [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+                if(gameboard.isFieldAdjacentToBomb(i, j) )
+                    std::cout << " " << gameboard.countAdjacentBombAt(i, j);
+                else 
+                    std::cout << " O";
+            }
+        }
+        std::cout << std::endl;
+    }
+    std::cout << std::endl;
+    usleep(10000);
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply src/Viewer.cpp]: done

File: tests/GameboardTest.cpp (hunk 1 above)
Comment (dariannax3) [reply tests/GameboardTest.cpp]: done

File: tests/GameboardTest.cpp [hunk 10]
```diff
@@ -20,4 +21,133 @@ TEST(RandomGeneratorTest, getRandomTest) {
...
+    void checkVisibilityOfAllFields(Visibility visibility) {
+        for (int row = 0; row < gameboard.getHeight(); row++) {
+            for (int column = 0; column < gameboard.getWidth(); column++) {
+                EXPECT_EQ(gameboard.getFieldAt(row, column).visibility, visibility);
+            }
+        }
+    }
+
+    void checkVisibilityOfSpecificRowsAndFields(int row_start, int row_end, int column_start, int column_end, Visibility visibility) {
+        if(isRangeCorrect(row_start, row_end, column_start, column_end)){
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply tests/GameboardTest.cpp]: done

File: tests/GameboardTest.cpp [hunk 11]
```diff
@@ -20,4 +21,133 @@ TEST(RandomGeneratorTest, getRandomTest) {
...
+        gameboard.flagField(x+1,y);
+        gameboard.flagField(x+1,y+1);
+        gameboard.flagField(x,y-1);
+        gameboard.flagField(x,y);
+        gameboard.flagField(x,y+1);
+    }
+
+};
+
+TEST_F(GameboardTestFixture, GivenCoveredBoardWhenUncoverAllFieldsThanAllFieldsVisibilityStatusShouldBeUncover) {
```
[the remark this answers is not in the document]
Comment (dariannax3) [reply tests/GameboardTest.cpp]: done

File: tests/GameboardTest.cpp (hunk 2 above)
Comment (dariannax3) [reply tests/GameboardTest.cpp:163]: done

File: tests/GameboardTest.cpp (hunk 3 above)
Comment (dariannax3) [reply tests/GameboardTest.cpp]: done

**Review by Rosein — APPROVED** [review]
Good job!

**Pull request merged**